component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordChangeHandler.java

```diff
@@ -0,0 +1,166 @@
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.policy.password;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.event.IdentityEventConstants;
+import org.wso2.carbon.identity.event.IdentityEventException;
+import org.wso2.carbon.identity.event.event.Event;
+import org.wso2.carbon.identity.event.handler.AbstractEventHandler;
+import org.wso2.carbon.identity.governance.IdentityGovernanceException;
+import org.wso2.carbon.identity.governance.common.IdentityConnectorConfig;
+import org.wso2.carbon.user.core.UserStoreException;
+import org.wso2.carbon.user.core.UserStoreManager;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+
+/**
+ * Event Handler class which handles password update by user, password update by admin and add user events.
+ * <p>
+ * This updates the http://wso2.org/claims/lastPasswordChangedTimestamp claim upon the password change.
+ * This also publishes the password change event to IS Analytics.
+ */
+public class PasswordChangeHandler extends AbstractEventHandler implements IdentityConnectorConfig {
+    private static final Log log = LogFactory.getLog(PasswordChangeHandler.class);
+
+    @Override
+    public void handleEvent(Event event) throws IdentityEventException {
+        // Fetching event properties
+        String username = (String) event.getEventProperties().get(IdentityEventConstants.EventProperty.USER_NAME);
+        UserStoreManager userStoreManager = (UserStoreManager) event.getEventProperties()
+                .get(IdentityEventConstants.EventProperty.USER_STORE_MANAGER);
+        long timestamp = System.currentTimeMillis();
+
+        // Updating the last password changed claim
+        Map<String, String> claimMap = new HashMap<>();
+        claimMap.put(PasswordPolicyConstants.LAST_CREDENTIAL_UPDATE_TIMESTAMP_CLAIM, Long.toString(timestamp));
+        try {
+            userStoreManager.setUserClaimValues(username, claimMap, null);
+            if (log.isDebugEnabled()) {
+                log.debug("The claim uri "
+                        + PasswordPolicyConstants.LAST_CREDENTIAL_UPDATE_TIMESTAMP_CLAIM + " of "
+                        + username + " updated with the current timestamp");
+            }
+        } catch (UserStoreException e) {
+            log.error("Failed to update claim value for "
```

Comment:
If you catch it and log it and rethrow it, then there's no way for the upstream code to know that you've already logged the exception, and so the same exception might get logged twice.

Reply:
Fixed.